design/cors-configuration.md

```diff
@@ -0,0 +1,188 @@
+
+This proposal introduces a way to set all the CORS related configuration in Contour, letting Envoy do all the heavy work.
+
+## High-Level Design
+
+Envoy supports [CORS via a filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http_filters/cors_filter) and it can be configured [using the API](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto?#route-corspolicy). The changes proposed in this document will allow the configuration of CORS policies in Contour.
+
+At a high level the proposed changes will imply:
+- Adding new fields at route match level to configure the CORS policy in the YAML.
+- Changing some structs in the code.
+- Enabling the Envoy CORS filter.
+
+### Proposed YAML fields
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata:
+ name: google
+ namespace: prod
+spec:
+ virtualhost:
+   fqdn: www.google.com
+   routes:
+   - match: /finance
+     # The CORS policy described here is local to match and apply for "/finance" only.
+     corsPolicy:
+       # Which domains can access the matched resources in a cross-site manner.
+       allowOrigin:
+         - "*"
+       # Which HTTP methods are allowed for cross-origin requests (case-sensitive).
+       allowMethods:
+         - GET
+         - POST
+       # The headers the server is going to accept (case-insensitive).
+       allowHeaders:
+         - cache-control
+         - content-type
+         - custom-header
+       # The non simple headers the client will be able to access (case-insensitive).
+       exposeHeaders:
+         - Content-Length
+         - Content-Range
+       # Whether the server allows sending credentials (cookies for instance) in cross-origin requests.
+       allowCredentials: true
+       # The number of seconds the preflight response can be cached.
+       maxAge: "1728000"
+     services:
+     - name: finance-app
+       port: 9999
+```
+
+The names and types of the proposed fields are inspired by the [HTTP headers](https://www.w3.org/TR/2010/WD-cors-20100727/#syntax). The `maxAge` is proposed as a string field type even though it's a delta-seconds. The motivation for this is that if we declare it as an `integer`, in cases where no value is provided, it  would get automatically the 0 default value. This is not what most people are going to expect and when not specified, it's better to apply the [browser defaults](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age#Directives).
+
+## Detailed Design
+
+### Reading from YAML
+The ingressroute spec struct will be updated to contain the CORS related information:
+
+```go
+// contour/apis/contour/v1beta1/ingressroute.go
+// Route contains the set of routes for a virtual host
+type Route struct {
+   [... other members ...]
+   // Specifies the CORS policy to apply for the matched path.
+    CorsPolicy *CorsPolicy `json:"corsPolicy,omitempty"`
```

Comment:
You are right, probably the most common use case would be to apply the same policy across a virtual host but I would love to get feedback from the community about this.

Reply:
Having used this fork, I can confirm that yes, it would be good to apply the same policy across a virtual host. It perhaps doesn't give as much flexibility, but it's also a complete pain duplicating the CORS setup for each route.

Reply:
Would this then make the CorsPolicy (maybe `OriginPolicy` is a more generic name) a peer to the TLS policy on the virtualhost struct? If so, I like that, because it gives one place to set this information - the root HTTPProxy, with the caveat that there should be a way to disable on nonroots.

Reply:
I am not confident that I understand the use cases that would not be possible if the policy is moved to the virtual host. Would be great if someone could explain to me in small words :)

Reply:
> Would this then make the CorsPolicy (maybe `OriginPolicy` is a more generic name) a peer to the TLS policy on the virtualhost struct?

Even though I agree that `OriginPolicy` is a more generic name, I think that it could confuse a lot of people that would be looking for the `Cors` term. Here is a chart from Google trends comparing the popularity of "Same Origin Policy" and "CORS".

![Captura de pantalla 2020-07-22 a las 18 29 23](https://user-images.githubusercontent.com/1999627/88202716-91433080-cc49-11ea-9a8a-1da487dee34f.png)


> If so, I like that, because it gives one place to set this information - the root HTTPProxy, with the caveat that there should be a way to disable on nonroots.

We could add a `disableCors` property at route level. WDYT?

Reply:
Adding a `disableCors` at route level is a plausible fix, year, but then we also need to consider if cluster admins may want to not allow that (basically, do we need a `disableDisableCors` setting at the root). Welcome to the administrative override design space!

I don't know enough about how this is usually used to know how important being able to disable CORS on some routes is versus the cluster admin's ability to ensure it's enforced everywhere.

Reply:
> I am not confident that I understand the use cases that would not be possible if the policy is moved to the virtual host. Would be great if someone could explain to me in small words :)

The only case I could think of is the following one:  you have a public facing API and therefore, you want to give access to everyone that wants to use it. For instance,  if your API is deployed at `https://www.googleapis.com`, you need to set a cors policy where the allowed origins are `*` (because any domain could make requests to it). In this case, setting the policy at host level would be enough. But what if for whatever reason you want to block cross origin requests for a specific path like `/internal-api`? In this case, you could keep the aforementioned cors policy at host level but disable it for the `/internal-api` route. 

IMHO, the route level cors policies are edge cases and we could avoid them initially.